library/sm3.c

```diff
@@ -0,0 +1,474 @@
+ */
+/*
+ *  GB/T 32905-2016 compliant SM3 implementation
+ *  The SM3 algorithm was designed by Xiaoyun Wang et al in 2010.
+ *
+ *  http://www.gmbz.org.cn/upload/2018-07-24/1532401392982079739.pdf
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_SM3_C)
+
+#include "mbedtls/sm3.h"
+#include "mbedtls/platform_util.h"
+#include "mbedtls/error.h"
+
+#include <string.h>
+
+#if defined(MBEDTLS_SELF_TEST)
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#define mbedtls_printf printf
+#endif /* MBEDTLS_PLATFORM_C */
+#endif /* MBEDTLS_SELF_TEST */
+
+#define SM3_VALIDATE_RET(cond)                           \
+    MBEDTLS_INTERNAL_VALIDATE_RET( cond, MBEDTLS_ERR_SM3_BAD_INPUT_DATA )
+#define SM3_VALIDATE(cond)  MBEDTLS_INTERNAL_VALIDATE( cond )
+
+#if !defined(MBEDTLS_SM3_ALT)
+
+/*
+ * 32-bit integer manipulation macros (big endian)
+ */
+#ifndef GET_UINT32_BE
+#define GET_UINT32_BE(n,b,i)                            \
+do {                                                    \
+    (n) = ( (uint32_t) (b)[(i)    ] << 24 )             \
+        | ( (uint32_t) (b)[(i) + 1] << 16 )             \
+        | ( (uint32_t) (b)[(i) + 2] <<  8 )             \
+        | ( (uint32_t) (b)[(i) + 3]       );            \
+} while( 0 )
+#endif
+
+#ifndef PUT_UINT32_BE
+#define PUT_UINT32_BE(n,b,i)                            \
+do {                                                    \
+    (b)[(i)    ] = (unsigned char) ( (n) >> 24 );       \
+    (b)[(i) + 1] = (unsigned char) ( (n) >> 16 );       \
+    (b)[(i) + 2] = (unsigned char) ( (n) >>  8 );       \
+    (b)[(i) + 3] = (unsigned char) ( (n)       );       \
+} while( 0 )
+#endif
+
+void mbedtls_sm3_init( mbedtls_sm3_context *ctx )
+{
+    SM3_VALIDATE( ctx != NULL );
+
+    memset( ctx, 0, sizeof( mbedtls_sm3_context ) );
+}
+
+void mbedtls_sm3_free( mbedtls_sm3_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_sm3_context ) );
+}
+
+void mbedtls_sm3_clone( mbedtls_sm3_context *dst,
+                        const mbedtls_sm3_context *src )
+{
+    SM3_VALIDATE( dst != NULL );
+    SM3_VALIDATE( src != NULL );
+
+    *dst = *src;
+}
+
+/*
+ * SM3 context setup
+ */
+int mbedtls_sm3_starts_ret( mbedtls_sm3_context *ctx )
+{
+    SM3_VALIDATE_RET( ctx != NULL );
+
+    ctx->total[0] = 0;
+    ctx->total[1] = 0;
+
+    /* initial IV */
+    ctx->state[0] = 0x7380166F;
+    ctx->state[1] = 0x4914B2B9;
+    ctx->state[2] = 0x172442D7;
+    ctx->state[3] = 0xDA8A0600;
+    ctx->state[4] = 0xA96F30BC;
+    ctx->state[5] = 0x163138AA;
+    ctx->state[6] = 0xE38DEE4D;
+    ctx->state[7] = 0xB0FB0E4E;
+
+    return( 0 );
+}
+
+#if !defined(MBEDTLS_SM3_PROCESS_ALT)
+#define EXTMSG_W1_LEN 68
+#define EXTMSG_W2_LEN 64
+#define EXTMSG_W_LEN (EXTMSG_W1_LEN + EXTMSG_W2_LEN)
+
+#define ROTL32( value, amount ) \
+    ( (uint32_t) ( (value) << (amount) ) | ( (value) >> ( 32 - (amount) ) ) )
```

Comment:
Change it to `( (uint32_t) ( (value) << (amount) ) | ( (value) >> ( (32 - (amount) )& 0x1f ) ) )`

That can fix [Travis CI error](https://travis-ci.com/github/ARMmbed/mbedtls/jobs/505827794#L1275). The error is a sanitize error. With `-fsanitize=undefined` option, it will report this error .